### PR TITLE
[SPARK-15684][SparkR]Not mask startsWith and endsWith in R

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

In R 3.3.0, startsWith and endsWith are added. In this PR, I make the two work in SparkR. 
1. Remove signature in generic.R
2. Add setMethod in column.R
3. Add unit tests
## How was this patch tested?

Manually test it through SparkR shell for both column data and string data, which are added into the unit test file.

### CODE REVIEW

**Comment by wangmiao1981** [comment]
@shivaram It seems that integration server uses a lower version of R. Do you know whether has a mechanism to optionally compiling functions by checking version? For example, in C, we can use #ifdef #endif to enable/disable code for specific versions.

**Comment by shivaram** [comment]
@wangmiao1981 the best way to do this is to check the R version using `R.Version()` (similar to what you did in the unit tests)

File: R/pkg/R/column.R [hunk 1]
```diff
@@ -151,6 +151,40 @@ setMethod("substr", signature(x = "Column"),
...
+#' startsWith
+#'
+#' Determines if entries of x start with string (entries of) prefix respectively,
+#' where strings are recycled to common lengths.
+#'
+#' @rdname startsWith
+#' @name startsWith
+#' @family colum_func
+#'
+#' @param vector of character string whose “starts” are considered
```
Comment (felixcheung) [inline R/pkg/R/column.R]: `@param` should be once per parameter. I think we should leave out `@param` on the next line since it is describing the same parameter.
Also we should name the parameter first at the beginning of the text right after thr `@param` tag, see L146 for an example.
Finally, add `@param` and description for `x`

File: R/pkg/R/column.R [hunk 2]
```diff
@@ -151,6 +151,40 @@ setMethod("substr", signature(x = "Column"),
...
+#' endsWith
+#'
+#' Determines if entries of x end with string (entries of) suffix respectively,
+#' where strings are recycled to common lengths.
+#'
+#' @rdname endsWith
+#' @name endsWith
+#' @family colum_func
+#'
+#' @param vector of character string whose “ends” are considered
```
Comment (felixcheung) [inline R/pkg/R/column.R]: ditto here

**Comment by felixcheung** [comment]
please remove the workaround in test_context.R

File: R/pkg/R/generics.R [hunk 3]
```diff
@@ -695,10 +695,6 @@ setGeneric("desc", function(x) { standardGeneric("desc") })
 
 #' @rdname column
 #' @export
-setGeneric("endsWith", function(x, ...) { standardGeneric("endsWith") })
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: you shouldn't need to remove it - I think if they are updated to match the signature of the existing generic (in newer R 3.3.0) it should work both in the case of R 3.3.0 and R 3.2.0.
With regards to the comment on R version check - it should not be necessary also.
@shivaram

Comment (wangmiao1981) [inline R/pkg/R/generics.R]: I kept this definition before. But sparkR complains no signature found. The tests only work if I remove it in generic.R. Let me try again.

Comment (felixcheung) [inline R/pkg/R/generics.R]: to be clear, try something like
```
setGeneric("endsWith", function(x, suffix) { standardGeneric("endsWith") })
```

File: R/pkg/R/column.R (hunk 1 above)
Comment (wangmiao1981) [inline R/pkg/R/column.R]: I will add param name. The first line is for `x` and the second line is for `prefix`

**Comment by wangmiao1981** [comment]
@felixcheung Any more comments? The version check is needed otherwise it can't pass unit tests. All other review comments are done.

File: R/pkg/R/generics.R [hunk 4]
```diff
@@ -691,11 +691,11 @@ setGeneric("contains", function(x, ...) { standardGeneric("contains") })
 
 #' @rdname column
 #' @export
-setGeneric("desc", function(x) { standardGeneric("desc") })
+setGeneric("endsWith", function(x, suffix) { standardGeneric("endsWith") })
```
Comment (shivaram) [inline R/pkg/R/generics.R]: can you move this back into the existing order ? it is sorted alphabetically if i'm not wrong.

File: R/pkg/R/generics.R [hunk 5]
```diff
@@ -723,11 +723,11 @@ setGeneric("like", function(x, ...) { standardGeneric("like") })
 
 #' @rdname column
 #' @export
-setGeneric("rlike", function(x, ...) { standardGeneric("rlike") })
+setGeneric("startsWith", function(x, prefix) { standardGeneric("startsWith") })
 
 #' @rdname column
 #' @export
-setGeneric("startsWith", function(x, ...) { standardGeneric("startsWith") })
+setGeneric("rlike", function(x, ...) { standardGeneric("rlike") })
```
Comment (shivaram) [inline R/pkg/R/generics.R]: same as above

File: R/pkg/inst/tests/testthat/test_sparkSQL.R [hunk 6]
```diff
@@ -1137,6 +1137,13 @@ test_that("string operators", {
   expect_equal(count(where(df, like(df$name, "A%"))), 1)
   expect_equal(count(where(df, startsWith(df$name, "A"))), 1)
   expect_equal(first(select(df, substr(df$name, 1, 2)))[[1]], "Mi")
+  if (as.numeric(R.version$major) >= 3 && as.numeric(R.version$minor) >= 3) {
```
Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: Do you know why this is needed? Jenkins is running with R 3.2~, not R 3.3.0, so this check has effectively disabled all the tests below.

Comment (felixcheung) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: actually, I think you should only put line 1144 and 1145 into the version check. startsWith(character) and endsWith(character) are not in R 3.2~ that's why the tests fail in Jenkins. The other with DataFrame/column should work though and we should test for it.

Comment (wangmiao1981) [inline R/pkg/inst/tests/testthat/test_sparkSQL.R]: I agree with you that it should work for Column anyway. These are all new tested added by me. I will move the column tests out.

**Comment by shivaram** [comment]
Thanks @wangmiao1981 - I think the updated change LGTM. @felixcheung Let me know if you have any more comments

**Comment by felixcheung** [comment]
LGTM!

**Comment by shivaram** [comment]
Merging this to master and branch-2.0
